core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp

```diff
@@ -85,16 +72,18 @@ class Kokkos::Impl::ParallelReduce<Functor, Kokkos::MDRangePolicy<Traits...>,
     }
 
     ValueType val;
-    typename Analysis::Reducer final_reducer(
-        ReducerConditional::select(m_functor, m_reducer));
-    final_reducer.init(&val);
+    const ReducerType& reducer = m_functor_reducer.get_reducer();
+    reducer.init(&val);
 
     Kokkos::Experimental::Impl::OpenACCParallelReduceMDRangeHelper(
-        Kokkos::Experimental::Impl::FunctorAdapter<Functor, Policy>(m_functor),
-        std::conditional_t<is_reducer_v<ReducerType>, ReducerType,
-                           Sum<ValueType>>(val),
+        Kokkos::Experimental::Impl::FunctorAdapter<FunctorType, Policy>(
+            m_functor_reducer.get_functor()),
+        std::conditional_t<
+            std::is_same_v<FunctorType, typename ReducerType::functor_type>,
+            Sum<ValueType>, typename ReducerType::functor_type>(val),
         m_policy);
 
+    reducer.final(&val);
```

Comment:
Why do we call `final` now and not before? Was there a bug?

Reply:
Yes, this was missing and I tried to enforce it multiple times. I thought this was a good opportunity but, of course, this is somewhat of a drive-by-change.